torch/csrc/jit/passes/onnx/eval_peephole.cpp

```diff
@@ -0,0 +1,154 @@
+  inputTensorValues.reserve(numInputs);
+  for (auto val : node->inputs()) {
+    if (val->node()->kind() == prim::Param) {
+      auto itr = valsToParamsMap.find(val);
+      if (itr == valsToParamsMap.end()) {
+        continue;
+      }
+      inputTensorValues.push_back(itr->second.second.toTensor());
+    } else if (val->node()->kind() == onnx::Constant) {
+      inputTensorValues.push_back(val->node()->t(attr::value));
+    } else {
+      continue;
+    }
+  }
+  return inputTensorValues;
+}
+
+// This pass fuses Conv and BatchNorm into Conv node
+// Conv and BatchNorm can be fused only if inputs for Batchnorm node:
+// scale, bias, mean and var are all tensors of same shape (C) and
+// if the size of the first dimension (dim 0) is the same between Conv
+// input weight and Batchnorm input scale
+static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      fuseConvBatchNorm(child_block, valsToParamsMap);
+    }
+    if (it->kind() == onnx::Conv) {
+      if (it->output()->uses().size() != 1) {
+        continue;
+      }
+      auto bnNode = it->output()->uses()[0].user;
+      if (bnNode->kind() != onnx::BatchNormalization) {
+        continue;
+      }
+      auto origconvNode = *it;
+      auto epsilon = bnNode->f(attr::epsilon);
+      auto w_conv_value = getValues(origconvNode, valsToParamsMap);
+      TORCH_INTERNAL_ASSERT(
+          w_conv_value.size() >= 1,
+          "convolution node expected to have at least one initializer");
+
+      auto bn_value = getValues(bnNode, valsToParamsMap);
+      TORCH_INTERNAL_ASSERT(
+          bn_value.size() == 4,
+          "batchnorm node expected to have four initializers");
+
+      auto bn_scale = bn_value[0].clone();
+      auto bn_B = bn_value[1].clone();
+      auto bn_mean = bn_value[2].clone();
+      auto bn_var = bn_value[3].clone();
+      auto w_conv = w_conv_value[0].clone();
+      at::Tensor b_conv;
+
+      if (!bn_scale.is_floating_point() || !bn_B.is_floating_point() ||
+          !bn_mean.is_floating_point() || !bn_var.is_floating_point() ||
+          !w_conv.is_floating_point() || bn_scale.dim() != 1 ||
+          bn_B.dim() != 1 || bn_mean.dim() != 1 || bn_var.dim() != 1 ||
+          !(bn_scale.size(0) == bn_B.size(0)) ||
+          !(bn_B.size(0) == bn_mean.size(0)) ||
+          !(bn_mean.size(0) == bn_var.size(0)) || !(w_conv.dim() > 2) ||
+          !(w_conv.size(0) == bn_scale.size(0))) {
+        continue;
+      }
+
+      bn_var = bn_var.add(epsilon);
+      bn_var = bn_var.sqrt();
+      bn_scale = bn_scale.div(bn_var);
+
+      // Calculate weight
+      for (size_t i = 0; i < w_conv.size(0); i++) {
+        w_conv[i] = w_conv[i].mul(bn_scale[i]);
+      }
+
+      // Calculate bias
+      if (origconvNode->inputs().size() == 3) {
+        b_conv = w_conv_value[1].clone();
```

Comment:
also within the scope of dynamic weight. This index may be out of scope if weight or bias is input.